File: cpp/valid_parenthesis_string.cpp
#include <iostream>
#include <string>
#include <stack>

class Solution {
public:
    bool checkValidString(std::string s) {
        if (s.length() == 0) return true;

        std::stack<int> left_stack, right_stack, star_stack;
        char current_char;

        // preprocessing
        for(int i = 0; i < s.length(); i++) {
            current_char = s[i];
            if (current_char == '(') left_stack.push(i);
            else if (current_char == ')') {
                if (left_stack.size() == 0 && star_stack.size() == 0) return false;
                if (left_stack.size()) left_stack.pop();
                else if (!left_stack.size() && star_stack.size()) right_stack.push(i);
            } else star_stack.push(i);
        }

        // post processing
        int idx, star_index;
        if (!left_stack.empty()) {
            
            while(!left_stack.empty() && !star_stack.empty()) {
                idx = left_stack.top();
                star_index = star_stack.top();
                star_stack.pop();
                if (idx < star_index) left_stack.pop();
            }
        }

        if (!right_stack.empty()) {
            while (!right_stack.empty() && !star_stack.empty()) {
                idx = right_stack.top();
                star_index = star_stack.top();
                star_stack.pop();
                if (star_index < idx) right_stack.pop();
            }
        }

        return left_stack.empty() && right_stack.empty();
    }
};

int main() {
    Solution solution = Solution();
    std::cout << solution.checkValidString("(())") << std::endl;

    return 0;
}

File: set_2/valid_parenthesis_string.py
import unittest

class Solution(object):

    def __init__(self): return

    def valid_parenthesis_string(self, string):
        if len(string) == 0: return True

        left_stack, star_stack, right_stack = [], [], []

        for index, char in enumerate(string):
            if char == '(': 
                left_stack.append((index, '('))
            if char == ')': 
                if len(left_stack) == 0 and len(star_stack) == 0: return False
                if len(left_stack): left_stack.pop()
                elif len(left_stack) == 0 and len(star_stack): right_stack.append((index, ')'))
            if char == '*':
                star_stack.append((index, '*'))
        
        # post processing
        if len(left_stack):
            while len(left_stack) and len(star_stack):
                left_index = left_stack[len(left_stack)-1][0]
                star_index, _ = star_stack.pop()
                if left_index < star_index: left_stack.pop()
        if len(right_stack):
            while len(right_stack) and len(star_stack):
                right_index = right_stack[len(right_stack)-1][0]
                star_index, _ = star_stack.pop()
                if star_index < right_index: right_stack.pop()
        return len(left_stack) == 0 and len(right_stack) == 0
    
class Test(unittest.TestCase):

    def test_all_stars(self):
        self.assertEqual(True, Solution().valid_parenthesis_string('*********'))

    def test_random_mix(self):
        self.assertEqual(True, Solution().valid_parenthesis_string('(*)*(()*'))

    def test_random(self):
        self.assertEqual(True, Solution().valid_parenthesis_string('*)(*'))

    def test_invalid_left(self):
        self.assertEqual(False, Solution().valid_parenthesis_string('(*('))
    
    def test_invalid_right(self):
        self.assertEqual(False, Solution().valid_parenthesis_string(')*)*'))

    def test_valid_long(self):
        self.assertEqual(False, Solution().valid_parenthesis_string("()(*(*()()(*)(()())((*))()(()(*((*)))))*)()(()))(("))

unittest.main(verbosity=2)
